File: Easy/9/9.cpp
#include <climits>
#include <iostream>

using namespace std;

// Reverse the number and compare it to the original to verify if it is a
// palindrome or not
class Solution {
public:
  bool isPalindrome(int x) {
    int num = x;
    if (x < 0)
      return false;
    int rev = 0;
    while (x != 0) {
      int digit = x % 10;
      if (rev > INT_MAX / 10 || (rev == INT_MAX / 10 && digit > 7))
        return false;
      if (rev < INT_MIN / 10 || (rev == INT_MIN / 10 && digit < -8))
        return false;
      rev = rev * 10 + digit;
      x = x / 10;
    }
    return num == rev;
  }
};

int main() {
  cout << "It's Showtime Folks!" << endl;
  cout << "Solution: " << Solution().isPalindrome(121) << endl;
  cout << "Solution: " << Solution().isPalindrome(-121) << endl;
  cout << "Solution: " << Solution().isPalindrome(10) << endl;
  cout << "Solution: " << Solution().isPalindrome(1234567899) << endl;
  return 0;
}

File: Easy/9/9.js
/**
 * @param {number}
 * @return {boolean}
 */
var isPalindrome = function (x) {
  if (x < 0) return false;
  const num = x;
  let rev = 0;
  while (x != 0) {
    digit = x % 10;
    // There is no need to verify that the number is within the 32-bit limit
    rev = rev * 10 + digit;
    x = Math.trunc(x / 10);
  }
  return num == rev;
};

console.log("It's Showtime Folks!");
console.log(`Solution: ${isPalindrome(121)}`);
console.log(`Solution: ${isPalindrome(-121)}`);
console.log(`Solution: ${isPalindrome(10)}`);
console.log(`Solution: ${isPalindrome(1234567899)}`);
